src/Wine.cc

```diff
@@ -0,0 +1,175 @@
+  };
+  map<string, library> libraries;
+
+  // iterate through the mappings and search for PE libraries
+  // see wine, packet_query_libraries_cb
+  char buf[0x400];
+  for (auto it = mapping_list.begin(); it != mapping_list.end(); it++) {
+    auto km = *it;
+    bool ok = true;
+    t->read_bytes_helper(km.start(), sizeof(buf), buf, &ok);
+    if (ok) {
+      struct IMAGE_DOS_HEADER* dosheader = (struct IMAGE_DOS_HEADER*)buf;
+      struct IMAGE_NT_HEADERS* ntheader = (struct IMAGE_NT_HEADERS*)(buf + dosheader->e_lfanew);
+      if (dosheader->e_magic == IMAGE_DOS_SIGNATURE &&
+          dosheader->e_lfanew < sizeof(buf)-4 &&
+          ntheader->Signature == IMAGE_NT_SIGNATURE)
+      {
+        if (ntheader->OptionalHeader.Magic == IMAGE_NT_OPTIONAL_HDR32_MAGIC ||
+            ntheader->OptionalHeader.Magic == IMAGE_NT_OPTIONAL_HDR64_MAGIC)
+        {
+          vector<remote_ptr<void>> segments;
+          struct IMAGE_SECTION_HEADER* sec = IMAGE_FIRST_SECTION(ntheader);
+          for (int i = 0; sec && i < std::max(ntheader->FileHeader.NumberOfSections, (uint16_t)1); ++i) {
+            struct IMAGE_SECTION_HEADER* seci = sec + i;
+            if ((char*)seci > buf + sizeof(buf)) break;
+            segments.push_back(km.start() + seci->VirtualAddress);
+          }
+
+          string name = km.fsname();
+
+          // If the mapping with the PE signature has no name assigned search the other segments for a name.
+          for (auto it2 = mapping_list.begin(); name.empty() && it2 != mapping_list.end(); it2++) {
+            if (find(segments.begin(), segments.end(), it2->start()) != segments.end()) {
+              if (!it2->fsname().empty()) {
+                name = it2->fsname();
+              }
+            }
+          }
+
+          // we found no name
+          if (name.empty()) {
+            continue;
+          }
+
+          // e.g. gdi32.so is prepared by wine to be found as PE image, but gdb does not like it.
+          if (name.substr(name.length() - 3, 3) == ".so") {
+            continue;
+          }
+
+          libraries[name].segments = segments;
+          libraries[name].has_exec = true;
+        }
+      }
+    }
+  }
+
+  // remove mappings related to PE libraries
+  for (auto m = mapping_list.begin(); m != mapping_list.end(); ) {
+    bool del = false;
+    for (auto lib : libraries) {
+      if (m->fsname() == lib.first) del = true;
+      if (!del) {
+        for (auto seg : lib.second.segments) {
+          if (seg == m->start()) del = true;
+        }
+      }
+    }
+    if (del)
+      m = mapping_list.erase(m);
+    else
+      m++;
+  }
+
+  // search remaining for regular shared objects
+  for (auto it = mapping_list.begin(); it != mapping_list.end(); it++) {
+    auto km = *it;
+    string name = km.fsname();
+    if (!name.empty()) {
+      libraries[name].segments.push_back(km.start());
+      if (km.prot() & PROT_EXEC) {
+        libraries[name].has_exec = true;
+      }
+    }
+  }
```

Comment:
What if the same library is mapped twice for some reason?

Reply:
If some code creates an executable memory mapping that's not a file (e.g. JITted code), that will show up in `libraries` I guess. What will gdb do then?

Reply:
> What if the same library is mapped twice for some reason?

Current version would send addresses from both within one library to gdb. I guess this is one such a "fragility". Have to investigate.

Reply:
> If some code creates an executable memory mapping that's not a file (e.g. JITted code), that will show up in `libraries` I guess. What will gdb do then?

Currently this will just show up if there is a filename in the mapping present.